DataProcessing as a top-level object and independent of Repertoire

schristley opened the pull request:
TODO: need to incorporate design decisions from the `DataProcessing` sprint.

--- this is old ---

Allowing only a single `DataProcessing` object for a `Repertoire` object implies:

- If the same set of sequencing files are data processed in multiple ways, e.g. MiXCR for one workflow and IgBlast for another, those require separate `Repertoire` objects with their own id.

- That single `DataProcessing` is implicitly the primary annotation, so there is no need for the `primary_annnotation` flag.

- Both `repertoire_id` and `data_processing_id` are not needed together to select the appropriate set of rearrangements for a Repertoire, now `repertoire_id` is sufficient. Thus, `data_processing_id` can be removed from `Rearrangement` unless it's needed there for another purpose.

- This makes the `Repertoire` concept more solid as being the annotated rearrangements for a biological repertoire for a single subject. A more abstract object, like `RepertoireSet` #445 , is needed to support flexible analysis.

This doesn't address other issues regarding the `DataProcessing` object.

schristley commented:
Additional changes would be needed to the validation routines, the blank repertoire template, etc., before this PR could be finalized.

scharch commented:
I'm sure this in the "etc," @schristley, but to keep @bussec happy let's make explicit that it should also included changes to the docs making clear how we expect `Repertoire` to be used... 😊

bcorrie commented:
Do we want to make this change before we sort out #441?

Making this change has a pretty dramatic follow on effect, in particular on repositories that currently might have/need a data_processing_id and repertoire_id to link rearrangements back to repertoires. This means that every record in a repository (both repertoire and rearrangement records) would need to be updated.

I would **hate** to do that process on the 2 Billion odd rearrangements that we have in our repositories and then we change our mind because we decide to change Repertoire again. I strongly suggest that we sort out how all of Repertoire, Rearrangement, Cell, Clone etc all fit together and interact with DataProcessing before we make a change with such a significant impact. This feels like a 2.0 change to me...

schristley commented:
> Do we want to make this change before we sort out #441?

Probably not, but if we get far enough along with #441 then it might be worthwhile to put this in with other changes to prevent the array from being used too extensively.

> Making this change has a pretty dramatic follow on effect, in particular on repositories that currently might have/need a data_processing_id and repertoire_id to link rearrangements back to repertoires. This means that every record in a repository (both repertoire and rearrangement records) would need to be updated.
> 
> I would **hate** to do that process on the 2 Billion odd rearrangements that we have in our repositories and then we change our mind because we decide to change Repertoire again.

Agreed. I suppose the good thing is that IPA only uses a single data processing, and same for VDJServer, so that covers a large percentage. There's an extra `data_processing_id` field in rearrangements, which isn't needed, but it isn't mandatory that it be removed either.

schristley commented:
One point to make is that even though this structure is fine for curation, when there is one final data processing that's released for publication, it doesn't support the discovery/exploration/analysis workflow as well.

For example, let's say I want to try different quality filters and other pre-processing parameters. I run my pre-processing tool on the raw sequencing files and get one set of output. Then I run the tool again with a different set of parameters and get another set of output. Let's say the tool is special, it knows about `Repertoire`, and in fact as input, you give it a list of `Repertoires` and that's how it knows which raw sequencing files to process. Furthermore, the tool knows about `DataProcessing` and automatically fills in the information in the `Repertoires` about how it was run, with its parameters and such. Every time I run the tool, it's going to replace the data processing with new info, and there really isn't a direct way to keep track of those multiple runs, for example to compare them.

I don't consider this a major issue, but it's something to think about if you are developing a tool that will _write_ metadata information versus just reading it.

scharch commented:
> Every time I run the tool, it's going to replace the data processing with new info

Interesting. The way I've been thinking about this for SONAR is actually updating, not overwriting, because SONAR is a suite of scripts and each step creates its own log entry already. So I actually end up with a different problem, which is how to "simplify" a huge agglomerated `data_processing` object to only the non-redundant steps at the end...

schristley commented:
> > Every time I run the tool, it's going to replace the data processing with new info
> 
> Interesting. The way I've been thinking about this for SONAR is actually updating, not overwriting, because SONAR is a suite of scripts and each step creates its own log entry already. So I actually end up with a different problem, which is how to "simplify" a huge agglomerated `data_processing` object to only the non-redundant steps at the end...

This is, in a way, why I added `analysis_provenance_id`, where from a VDJServer perspective, instead of filling up the `DataProcessing` object with lots of VDJServer-specific stuff, I can keep it in a separate object, referenced by that id. Now the spec says it points to a [PROV](https://www.w3.org/TR/2013/NOTE-prov-overview-20130430/) model (sadly the [python](https://prov.readthedocs.io/en/latest/) package is kinda crappy), but I don't think that's something we've discussed with any depth.

Imagine this, `DataProcessing` has this indicating that 3 tools were run:

```
software_version: "vdj_pipe-ls5-0.1.7u8 igblast-ls5-1.14u2 sonar-ls5-4.2u1"
```

The PROV model has then, at least, 3 high-level activities, for the 3 tools that were run. Here's might how I would record the provenance for running IgBlast:

```
document = prov.ProvDocument()

# an entity for the input file
e1 = document.entity('e1', (
    (prov.PROV_TYPE, "File"),
    ('ex:path', "my_sequences.fasta"),
    ('ex:creator', "Alice"),
))

# an entity for the output file
e2 = document.entity('e2', (
    (prov.PROV_TYPE, "File"),
    ('ex:path', "rearrangements.tsv"),
    ('ex:creator', "Alice"),
))

# create an activity for the igblast tool
a1 = document.activity('igblast-ls5-1.14u2', datetime.datetime.now(), None, None)

# indicate output file was derived from input file by the activity
document.wasDerivedFrom(e2, e1, a1)
```

Now in theory, this provides analysis provenance in a standard data model, and you could get as detailed and fine-grained as you want, or stay pretty high level. The resultant PROV document can be saved as JSON.

There is a [document](https://docs.google.com/document/d/1DLMUFPGrjoQCE4XsK26Bzu72Ig0-bTMI2PMlQhuhKoE/edit#heading=h.hsojms563y7g) that I wrote awhile back with my thoughts about analysis provenance.

The provenance that we maintain in VDJServer is hard-coded and non-standard, so it's still a long-time feature request to put in all in the PROV data model.

The question is, should AIRR try adopting this more formally for tools that want to _update_ an ongoing data processing with it's activities?

scharch commented:
Are you suggesting "replacing" `data_processing` with a PROV document? I could see that as a long-term enhancement...

schristley commented:
> Are you suggesting "replacing" `data_processing` with a PROV document? I could see that as a long-term enhancement...

No, only that it references a PROV document with `analysis_provenance_id`. I see `data_processing` as useful to provide some summary information.

javh reviewed (approved):
Fine by me.

bussec commented:
@schristley The proposed change seems reasonable to me, but this PR has been sitting here for a while and #441 was also closed a while ago. Did we discuss this more recently than 2020?

schristley commented:
There's been extensive discussion (and design) around `DataProcessing` as a top-level object. However, there's been minimal implementation so far. This PR predates the decision for `DataProcessing` as a top-level object, and essentially represents an alternative design but it that has some flaws, so I don't want to commit it. I didn't delete because I didn't want to forget my comments with @scharch . Instead, I've changed to draft, changed the milestone to V2, and changed the title to reflect the new design.

javh commented:
Thanks, @schristley. One more PR down. :)

javh commented:
From the call:
* Decide if this is still relevant. Last commit was 2020.
* Will leave open for discussion purposes.

bcorrie commented:
@schristley should we move this to v2.1 with other DataProcessing changes?

schristley commented:
Cleared milestone as its unknown when/if data processing will be re-designed.

javh commented:
> Cleared milestone as its unknown when/if data processing will be re-designed.

Seems highly unlikely we end up redesigning DataProcess at this point. I'm inclined to let this one fade.

javh commented:
@schristley Are you okay with us closing this?

schristley commented:
stale
